src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs

```diff
@@ -175,225 +138,134 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
                 return;
             }
 
-            var declarationStatement = (StatementSyntax)declarator.Parent.Parent;
+            // Check if the as operand is ever written up to the point of null check.
+            //
+            //      var s = field as string;
+            //      field = null;
+            //      if (s != null) { ... }
+            //
+            // It's no longer safe to use pattern-matching because 'field is string s' would never be true.
+            var asOperand = semanticModel.GetSymbolInfo(asExpression.Left, cancellationToken).Symbol;
+            var symbolName = asOperand?.Name;
+            if (symbolName != null)
+            {
+                var localStatementStart = localStatement.SpanStart;
+                var comparisonSpanStart = comparison.SpanStart;
+
+                foreach (var descendentNode in enclosingBlock.DescendantNodes())
+                {
+                    var descendentNodeSpanStart = descendentNode.SpanStart;
+                    if (descendentNodeSpanStart <= localStatementStart)
+                    {
+                        continue;
+                    }
+
+                    if (descendentNodeSpanStart >= comparisonSpanStart)
+                    {
+                        break;
+                    }
 
-            // If we convert this to 'if (o is Type x)' then 'x' will not be definitely assigned
-            // in the Else branch of the IfStatement, or after the IfStatement. Make sure
-            // that doesn't cause definite assignment issues.
-            if (!CheckDefiniteAssignment(semanticModel, localSymbol,
-                    declarationStatement, targetStatement, parentBlock, isNegativeNullCheck))
+                    if (descendentNode.IsKind(SyntaxKind.IdentifierName, out IdentifierNameSyntax identifierName) &&
+                        identifierName.Identifier.ValueText == symbolName &&
+                        asOperand.Equals(semanticModel.GetSymbolInfo(identifierName, cancellationToken).Symbol) &&
+                        identifierName.IsOnlyWrittenTo())
+                    {
+                        return;
+                    }
+                }
+            }
+
+            var analyzer = new Analyzer(
+                semanticModel,
+                localSymbol,
+                comparison,
+                operand,
+                localStatement,
+                enclosingBlock,
+                cancellationToken);
+            if (!analyzer.CanSafelyConvertToPatternMatching())
```

Comment:
>CanSafelyConvertToPatternMatching [](start = 26, length = 33)

Consider making this a static method of `Analyzer`, passing all the parameters necessary for construction. That method would instantiate the type and do the verification. The caller doesn't need to hold onto the instance.

Reply:
We need to use these parameters across all methods in `Analyzer`. To avoid passing all these to each method separately, we create a struct instead and pass those to the constructor once.

Reply:
What I'm suggesting is the same pattern we use for `LocalRewriter.Rewrite`:
```C#
struct Analyzer
{
    public static bool Analyze(... all the parameters ...)
    {
        var analyzer = new Analyzer(....); // call private constructor
        return analyzer.Analyze();
    }
    ... instance methods that you have today ...
}
```
(the method name, "Analyze", is just an example)